kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+
+
+-- Constants and default values
+
+local PREFIX = "[dns_client] "
+
+local DEFAULT_ERROR_TTL = 1     -- unit: second
+local DEFAULT_STALE_TTL = 4
+-- long-lasting TTL of 10 years for hosts or static IP addresses in cache settings
+local LONG_LASTING_TTL = 10 * 365 * 24 * 60 * 60
+
+local DEFAULT_FAMILY = { "SRV", "A", "AAAA" }
+
+local TYPE_SRV = resolver.TYPE_SRV
+local TYPE_A = resolver.TYPE_A
+local TYPE_AAAA = resolver.TYPE_AAAA
+local TYPE_A_AAAA = -1  -- used to resolve IP addresses for SRV targets
+
+local TYPE_TO_NAME = {
+  [TYPE_SRV] = "SRV",
+  [TYPE_A] = "A",
+  [TYPE_AAAA] = "AAAA",
+  [TYPE_A_AAAA] = "A/AAAA",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local HIT_LEVEL_TO_NAME = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "miss",
+  [4] = "hit_stale",
+}
+
+-- client specific error
+local CACHE_ONLY_ERROR_CODE = 100
+local CACHE_ONLY_ERROR_MESSAGE = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = {
+  errcode = CACHE_ONLY_ERROR_CODE,
+  errstr = CACHE_ONLY_ERROR_MESSAGE,
+}
+
+local EMPTY_RECORD_ERROR_CODE = 101
+local EMPTY_RECORD_ERROR_MESSAGE = "empty record received"
+
+
+-- APIs
+
+local _M = {
+  TYPE_SRV = TYPE_SRV,
+  TYPE_A = TYPE_A,
+  TYPE_AAAA = TYPE_AAAA,
+}
+local _MT = { __index = _M, }
+
+
+local _TRIES_MT = { __tostring = cjson.encode, }
+
+
+local function stats_init_name(stats, name)
+  if not stats[name] then
+    stats[name] = {}
+  end
+end
+
+
+local function stats_increment(stats, name, key)
+  stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+local function stats_set_count(stats, name, key, value)
+  stats[name][key] = value
+end
+
+
+local init_hosts do
+  local function insert_answer_into_cache(cache, hosts_cache, address, name, qtype)
+    local answers = {
+      ttl = LONG_LASTING_TTL,
+      expire = now() + LONG_LASTING_TTL,
+      {
+        name = name,
+        type = qtype,
+        address = address,
+        class = 1,
+        ttl = LONG_LASTING_TTL,
+      },
+    }
+
+    hosts_cache[name .. ":" .. qtype] = answers
+    hosts_cache[name .. ":" .. TYPE_A_AAAA] = answers
+  end
+
+  -- insert hosts into cache
+  function init_hosts(cache, path)
+    local hosts = parse_hosts(path)
+    local hosts_cache = {}
+
+    for name, address in pairs(hosts) do
+      name = string_lower(name)
+
+      if address.ipv6 then
+        insert_answer_into_cache(cache, hosts_cache, address.ipv6, name, TYPE_AAAA)
+      end
+
+      if address.ipv4 then
+        insert_answer_into_cache(cache, hosts_cache, address.ipv4, name, TYPE_A)
+      end
+    end
+
+    return hosts, hosts_cache
+  end
+end
+
+
+-- distinguish the worker_events sources registered by different new() instances
+local ipc_counter = 0
+
+function _M.new(opts)
+  opts = opts or {}
+
+  local enable_ipv4, enable_ipv6, enable_srv
+
+  for _, typstr in ipairs(opts.family or DEFAULT_FAMILY) do
+    typstr = typstr:upper()
+
+    if typstr == "A" then
```

Comment:
should we use `TYPE_TO_NAME[xxx]`?

Reply:
It's already string type, and only 3 types supported currently, A/AAAA/SRV, no LAST and CNAME, so I think its ok to compare them simply.